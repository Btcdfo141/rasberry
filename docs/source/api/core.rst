.. _core_module:

:mod:`spencerassistant.core`
-------------------------

.. automodule:: spencerassistant.core
    :members: