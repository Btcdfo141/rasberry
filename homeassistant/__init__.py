"""Init file for Home Assistant."""


test2
