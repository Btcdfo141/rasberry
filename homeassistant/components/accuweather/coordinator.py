"""The AccuWeather coordinator."""

from asyncio import timeout
from datetime import timedelta
import logging
from typing import TYPE_CHECKING, Any

from accuweather import AccuWeather, ApiError, InvalidApiKeyError, RequestsExceededError
from aiohttp.client_exceptions import ClientConnectorError

from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
from homeassistant.helpers.update_coordinator import (
    TimestampDataUpdateCoordinator,
    UpdateFailed,
)

from .const import DOMAIN, MANUFACTURER

EXCEPTIONS = (ApiError, ClientConnectorError, InvalidApiKeyError, RequestsExceededError)

_LOGGER = logging.getLogger(__name__)


class AccuWeatherDataUpdateCoordinator(TimestampDataUpdateCoordinator[Any]):
    """Class to manage fetching AccuWeather data API."""

    def __init__(
        self,
        hass: HomeAssistant,
        accuweather: AccuWeather,
        name: str,
        coordinator_type: str,
        update_interval: timedelta,
    ) -> None:
        """Initialize."""
        self.accuweather = accuweather

        if TYPE_CHECKING:
            assert accuweather.location_key is not None

        self.location_key = accuweather.location_key
        self.device_info = DeviceInfo(
            entry_type=DeviceEntryType.SERVICE,
            identifiers={(DOMAIN, self.location_key)},
            manufacturer=MANUFACTURER,
            name=name,
            # You don't need to provide specific details for the URL,
            # so passing in _ characters is fine if the location key
            # is correct
            configuration_url=(
                "http://accuweather.com/en/"
                f"_/_/{self.location_key}/"
                f"weather-forecast/{self.location_key}/"
            ),
        )

        super().__init__(
            hass,
            _LOGGER,
            name=f"{name} ({coordinator_type})",
            update_interval=update_interval,
        )


class AccuWeatherObservationDataUpdateCoordinator(AccuWeatherDataUpdateCoordinator):
    """Class to manage fetching AccuWeather data API."""

    async def _async_update_data(self) -> dict[str, Any]:
        """Update data via library."""
        try:
            async with timeout(10):
                result = await self.accuweather.async_get_current_conditions()
        except EXCEPTIONS as error:
            raise UpdateFailed(error) from error

        _LOGGER.debug("Requests remaining: %d", self.accuweather.requests_remaining)

        return result


class AccuWeatherDailyForecastDataUpdateCoordinator(AccuWeatherDataUpdateCoordinator):
    """Class to manage fetching AccuWeather data API."""

    async def _async_update_data(self) -> list[dict[str, Any]]:
        """Update data via library."""
        try:
            async with timeout(10):
                result = await self.accuweather.async_get_daily_forecast()
        except EXCEPTIONS as error:
            raise UpdateFailed(error) from error

        _LOGGER.debug("Requests remaining: %d", self.accuweather.requests_remaining)

        return result
