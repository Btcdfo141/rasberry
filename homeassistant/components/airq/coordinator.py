"""The air-Q integration."""
from __future__ import annotations

from datetime import timedelta
import logging

from aioairq import AirQ

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_IP_ADDRESS, CONF_PASSWORD, CONF_SCAN_INTERVAL
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import (
    CONF_CLIP_NEGATIVE,
    CONF_RETURN_AVERAGE,
    DOMAIN,
    MANUFACTURER,
    UPDATE_INTERVAL,
)

_LOGGER = logging.getLogger(__name__)

DEFAULT_OPTIONS = {
    CONF_SCAN_INTERVAL: UPDATE_INTERVAL,
    CONF_CLIP_NEGATIVE: True,
    CONF_RETURN_AVERAGE: True,
}


class AirQCoordinator(DataUpdateCoordinator):
    """Coordinator is responsible for querying the device at a specified route."""

    def __init__(
        self,
        hass: HomeAssistant,
        entry: ConfigEntry,
    ) -> None:
        """Initialise a custom coordinator."""
        self.options = DEFAULT_OPTIONS | entry.options
        super().__init__(
            hass,
            _LOGGER,
            name=DOMAIN,
            update_interval=timedelta(seconds=self.options[CONF_SCAN_INTERVAL]),
        )
        session = async_get_clientsession(hass)
        self.airq = AirQ(
            entry.data[CONF_IP_ADDRESS], entry.data[CONF_PASSWORD], session
        )
        self.device_id = entry.unique_id
        assert self.device_id is not None
        self.device_info = DeviceInfo(
            manufacturer=MANUFACTURER,
            identifiers={(DOMAIN, self.device_id)},
        )

    async def _async_update_data(self) -> dict:
        """Fetch the data from the device."""
        if "name" not in self.device_info:
            info = await self.airq.fetch_device_info()
            self.device_info.update(
                DeviceInfo(
                    name=info["name"],
                    model=info["model"],
                    sw_version=info["sw_version"],
                    hw_version=info["hw_version"],
                )
            )
        return await self.airq.get_latest_data(
            return_average=self.options[CONF_RETURN_AVERAGE],
            clip_negative_values=self.options[CONF_CLIP_NEGATIVE],
        )

    async def async_set_options(
        self, hass: HomeAssistant, config_entry: ConfigEntry
    ) -> None:
        """Update the configuration options for the device."""
        options = self.options | config_entry.options
        _LOGGER.debug(
            "%s.async_set_options: from %s to %s",
            self.__class__.__name__,
            repr(self.options),
            repr(options),
        )
        self.options = options
        self.update_interval = timedelta(seconds=self.options[CONF_SCAN_INTERVAL])
