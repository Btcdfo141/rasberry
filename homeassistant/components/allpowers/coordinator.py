"""Data coordinator for receiving Allpowers updates."""

import logging

from allpowers_ble import AllpowersBLE, models

from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


class AllpowersBLECoordinator(DataUpdateCoordinator):
    """Data coordinator for receiving Allpowers updates."""

    def __init__(self, hass: HomeAssistant, allpowers_ble: AllpowersBLE) -> None:
        """Initialise the coordinator."""
        super().__init__(
            hass,
            _LOGGER,
            name=DOMAIN,
        )
        self._allpowers_ble = allpowers_ble
        allpowers_ble.register_callback(self._async_handle_update)
        allpowers_ble.register_disconnected_callback(self._async_handle_disconnect)
        self.connected = False

    @callback
    def _async_handle_update(self, state: models.AllpowersState) -> None:
        """Just trigger the callbacks."""
        self.connected = True
        self.async_set_updated_data(True)

    @callback
    def _async_handle_disconnect(self) -> None:
        """Trigger the callbacks for disconnected."""
        self.connected = False
        self.async_update_listeners()
