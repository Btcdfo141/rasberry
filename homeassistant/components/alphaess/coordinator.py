"""Coordinator for AlphaEss integration."""

import logging

import aiohttp
from alphaess import alphaess

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import DOMAIN, SCAN_INTERVAL

_LOGGER: logging.Logger = logging.getLogger(__package__)


class AlphaESSDataUpdateCoordinator(DataUpdateCoordinator):
    """Class to manage fetching data from the API."""

    def __init__(self, hass: HomeAssistant, client: alphaess.alphaess) -> None:
        """Initialize."""
        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
        self.api = client
        self.update_method = self._async_update_data

    async def _async_update_data(self):
        """Update data via library."""
        try:
            _LOGGER.info("Trying to query AlphaESS api data")
            return await self.api.getdata()
        except (
            aiohttp.client_exceptions.ClientConnectorError,
            aiohttp.ClientResponseError,
        ) as error:
            raise UpdateFailed(error) from error
