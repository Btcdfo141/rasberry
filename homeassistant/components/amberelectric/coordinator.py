"""Amber Electric Coordinator."""
from __future__ import annotations

from datetime import timedelta

from amberelectric import ApiException
from amberelectric.api import amber_api
from amberelectric.model.actual_interval import ActualInterval
from amberelectric.model.channel import ChannelType
from amberelectric.model.current_interval import CurrentInterval
from amberelectric.model.forecast_interval import ForecastInterval
from amberelectric.model.site import Site

from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import LOGGER


def is_current(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
    """Return true if the supplied interval is a CurrentInterval."""
    return interval.__class__ == CurrentInterval


def is_forecast(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
    """Return true if the supplied interval is a ForecastInterval."""
    return interval.__class__ == ForecastInterval


def is_general(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
    """Return true if the supplied interval is on the general channel."""
    return interval.channel_type == ChannelType.GENERAL


def is_controlled_load(
    interval: ActualInterval | CurrentInterval | ForecastInterval,
) -> bool:
    """Return true if the supplied interval is on the controlled load channel."""
    return interval.channel_type == ChannelType.CONTROLLED_LOAD


def is_feed_in(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
    """Return true if the supplied interval is on the feed in channel."""
    return interval.channel_type == ChannelType.FEED_IN


def first(
    intervals: ActualInterval | CurrentInterval | ForecastInterval,
) -> ActualInterval | CurrentInterval | ForecastInterval | None:
    """Return the first element if it exists, otherwise return None."""
    if len(intervals) > 0:
        return intervals[0]
    return None


class AmberDataService:
    """AmberDataService - In charge of downloading the data for a site, which all the sensors read."""

    def __init__(
        self, hass: HomeAssistant, api: amber_api.AmberApi, site_id: str
    ) -> None:
        """Initialise the data service."""
        self._hass = hass
        self._api = api
        self._site_id = site_id
        self.coordinator: DataUpdateCoordinator | None = None

        self.data: list[ActualInterval | CurrentInterval | ForecastInterval]

        self.site: Site | None = None
        self.current_prices: dict[str, CurrentInterval | None] = {
            ChannelType.GENERAL: None,
            ChannelType.CONTROLLED_LOAD: None,
            ChannelType.FEED_IN: None,
        }

        self.forecasts: dict[str, list[ForecastInterval] | None] = {
            ChannelType.GENERAL: None,
            ChannelType.CONTROLLED_LOAD: None,
            ChannelType.FEED_IN: None,
        }

    @callback
    def async_setup(self) -> None:
        """Set up the data service."""
        self.coordinator = DataUpdateCoordinator(
            self._hass,
            LOGGER,
            name="amberelectric",
            update_method=self.async_update_data,
            update_interval=self.update_interval,
        )

    @property
    def update_interval(self) -> timedelta:
        """Update interval property."""
        return timedelta(minutes=1)

    def update(self) -> None:
        """Update callback."""
        try:
            sites = list(
                filter(lambda site: site.id == self._site_id, self._api.get_sites())
            )
            if len(sites) > 0:
                self.site = sites[0]

            self.data = self._api.get_current_price(self._site_id, next=48)
            self.current_prices = {
                ChannelType.GENERAL: None,
                ChannelType.CONTROLLED_LOAD: None,
                ChannelType.FEED_IN: None,
            }

            self.forecasts = {
                ChannelType.GENERAL: None,
                ChannelType.CONTROLLED_LOAD: None,
                ChannelType.FEED_IN: None,
            }

            current = list(filter(is_current, self.data))
            forecasts = list(filter(is_forecast, self.data))

            self.current_prices[ChannelType.GENERAL] = first(
                list(filter(is_general, current))
            )

            self.current_prices[ChannelType.CONTROLLED_LOAD] = first(
                list(filter(is_controlled_load, current))
            )

            self.current_prices[ChannelType.FEED_IN] = first(
                list(filter(is_feed_in, current))
            )

            self.forecasts[ChannelType.GENERAL] = list(filter(is_general, forecasts))
            self.forecasts[ChannelType.CONTROLLED_LOAD] = list(
                filter(is_controlled_load, forecasts)
            )
            self.forecasts[ChannelType.FEED_IN] = list(filter(is_feed_in, forecasts))

            LOGGER.debug("Fetched new Amber data: %s", self.data)

        except ApiException as api_exception:
            raise UpdateFailed("Missing price data, skipping update") from api_exception

    async def async_update_data(self) -> None:
        """Async update wrapper."""
        await self._hass.async_add_executor_job(self.update)
