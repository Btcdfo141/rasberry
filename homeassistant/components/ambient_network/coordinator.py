"""DataUpdateCoordinator for the Ambient Weather Network integration."""

from __future__ import annotations

from datetime import datetime, timedelta
from typing import Any, cast

from aioambient import OpenAPI
from aioambient.errors import RequestError

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_MAC
from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import API_LAST_DATA, DOMAIN, LOGGER, SCAN_INTERVAL


class AmbientNetworkDataUpdateCoordinator(DataUpdateCoordinator[dict[str, Any]]):
    """The Ambient Network Data Update Coordinator."""

    config_entry: ConfigEntry

    def __init__(self, hass: HomeAssistant, api: OpenAPI) -> None:
        """Initialize the coordinator."""
        super().__init__(hass, LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
        self.api = api
        self.data = {}
        self.outstanding_errors: list[str] = []

    def _report_error(self, message: str) -> None:
        """Report an error by logging it. Only log the first occurrence of each error."""

        if message not in self.outstanding_errors:
            LOGGER.warning(message)
            self.outstanding_errors.append(message)

    def _clear_errors(self) -> None:
        """Clear all outstanding errors."""

        if self.outstanding_errors:
            LOGGER.info(f"Station '{self.config_entry.title}' is back online")
            self.outstanding_errors = []

    async def _async_update_data(self) -> dict[str, Any]:
        """Fetch the latest data from the Ambient Network."""

        try:
            response = await self.api.get_device_details(
                self.config_entry.data[CONF_MAC]
            )
        except RequestError:
            self._report_error("Cannot connect to Ambient Network")
            return {}

        if (last_data := response.get(API_LAST_DATA)) is None:
            # Use previous data
            self._report_error(
                f"Station '{self.config_entry.title}' did not report any data"
            )
            last_data = self.data

        # Eliminate data if the station hasn't been updated for a while.
        if (created_at := last_data.get("created_at")) is None:
            self._report_error(
                f"Station '{self.config_entry.title}' did not report a time stamp"
            )
            return {}

        # Eliminate data that has been generated more than an hour ago. The station is
        # probably offline.
        if int(created_at / 1000) < int(
            (datetime.now() - timedelta(hours=1)).timestamp()
        ):
            self._report_error(
                f"Station '{self.config_entry.title}' reported stale data"
            )
            return {}

        self._clear_errors()
        return cast(dict[str, Any], last_data)
