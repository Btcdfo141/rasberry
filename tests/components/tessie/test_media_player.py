"""Test the Tessie media player platform."""

from datetime import timedelta

from freezegun.api import FrozenDateTimeFactory
from syrupy import SnapshotAssertion

from homeassistant.components.tessie.coordinator import TESSIE_SYNC_INTERVAL
from homeassistant.core import HomeAssistant

from .common import setup_platform

from tests.common import async_fire_time_changed

WAIT = timedelta(seconds=TESSIE_SYNC_INTERVAL)


async def test_media_player(
    hass: HomeAssistant, freezer: FrozenDateTimeFactory, snapshot: SnapshotAssertion
) -> None:
    """Tests that the media player entity is correct when idle."""

    assert len(hass.states.async_all("media_player")) == 0

    await setup_platform(hass)

    assert hass.states.async_all("media_player") == snapshot(name="idle")

    # Trigger coordinator refresh since it has a different fixture.
    freezer.tick(WAIT)
    async_fire_time_changed(hass)

    assert hass.states.async_all("media_player") == snapshot(name="playing")
