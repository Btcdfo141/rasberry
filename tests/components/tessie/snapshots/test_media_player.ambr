# serializer version: 1
# name: test_media_player[idle]
  list([
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'speaker',
        'friendly_name': 'Test Media player',
        'supported_features': <MediaPlayerEntityFeature: 0>,
        'volume_level': 0.22580323309042688,
      }),
      'context': <ANY>,
      'entity_id': 'media_player.test_media_player',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': 'idle',
    }),
  ])
# ---
# name: test_media_player[playing]
  list([
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'speaker',
        'friendly_name': 'Test Media player',
        'supported_features': <MediaPlayerEntityFeature: 0>,
        'volume_level': 0.22580323309042688,
      }),
      'context': <ANY>,
      'entity_id': 'media_player.test_media_player',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': 'idle',
    }),
  ])
# ---
